src/stdio_mgr/stdio_mgr.py

```diff
@@ -27,14 +27,55 @@
 """
 
 import sys
-from contextlib import contextmanager
-from io import BufferedReader, BytesIO, StringIO, TextIOBase, TextIOWrapper
+from contextlib import contextmanager, ExitStack, suppress
+from io import BufferedRandom, BufferedReader, BytesIO, TextIOBase, TextIOWrapper
 
 import attr
 
 
+class _SafeCloseIOBase(TextIOBase):
+    """Class to ignore ValueError during close.
+
+    Subclass of :cls:`~io.TextIOBase` that disregards ValueError which can
+    occur if the file has already been closed.
+    """
+
+    def close(self):
+        """Close and suppress ValueError.
+
+        ValueError may occur when the underlying
+        buffer is detached or the file was closed.
+        """
+        with suppress(ValueError):
+            super().close()
+
+
+class RandomTextIO(_SafeCloseIOBase, TextIOWrapper):
+    """Class to capture writes to a buffer even when detached.
+
+    Subclass of :cls:`~io.TextIOWrapper` that utlises an internal
+    buffer defaulting to utf-8 encoding.
+
+    All writes are flushed to the buffer.
+    """
+
+    def __init__(self):
+        """Initialise buffer with utf-8 encoding."""
+        self._buf = BufferedRandom(BytesIO())
+        super().__init__(self._buf, encoding="utf-8")
+
+    def write(self, *args, **kwargs):
+        """Flush after each write."""
+        super().write(*args, **kwargs)
+        self.flush()
+
+    def getvalue(self):
+        """Obtain buffer of text sent to the stream."""
+        return self._buf.raw.getvalue().decode(self.encoding)
+
+
 @attr.s(slots=False)
-class TeeStdin(TextIOWrapper):
+class TeeStdin(_SafeCloseIOBase, TextIOWrapper):
```

Comment:
Same as above, I think the [default encoding](https://github.com/bskinn/stdio-mgr/pull/26/files#diff-abadc7d46c647e14e338ebfbc175cc05R1270) should probably be `None`.

If changed, the [docstring](https://github.com/bskinn/stdio-mgr/pull/26/files#diff-abadc7d46c647e14e338ebfbc175cc05R120) would need to be changed, also.

Reply:
I'm going to look at adding the encoding to the `RandomTextIO`, but I would prefer to also default it to utf-8 here, and create a few tests.  The interface that I am replicating did not have an option to do different encodings - these encodings are for the internal buffers, which can not have a `None` encoding.  The amount of logic needed to have a `None` value on the external interface and internally use an appropriate default like Python does is not worth the effort at this stage.  And a serious multi-OS CI rig is needed to be testing how it behaves with different PYTHONIOENCODING and LANG/etc and other factors.

Reply:
Hmm... but if the effort here is to replicate `sys.stdfoo`, per [here](https://docs.python.org/3/library/sys.html#sys.stdout), "The character encoding is platform-dependent."

Which interface are you replicating, if not the actual `stdfoo`?

(This is not to claim that my previous implementation was at *all* correct, btw -- it was likely extremely kludge-tastic.)

Reply:
Oh, what you are suggesting definitely needs to be done eventually, but test suites rarely need it, and it wasn't previously supported, so there is little motivation to do it, compared with the other features in the issues list.  The feature of this PR and the previous one for stdin is to implement the correct class, especially to have the `.buffer`, which allows simulation of pipes for users who are peeking and proding at the buffer.

> and platform-dependent encoding for the portions exposed to the OS?

The key difference here is that this library does not expose these buffers to the OS -- the input is fake, and there is no real console - it is the console *application* which dictates the stdout/stderr encoding needs, otherwise it will render the stream as garbage.  We dont have that problem, and we are wasting our time doing a half-cocked version of a mock for those because the majority of console apps which need tests for those will still break under this wrapper.  To be anything like a real console, the library would need to implement `isatty`, and have a very good test suite.  And it would still be sub-par unless we mocked things like [`fileno`](https://docs.python.org/3/library/io.html#io.IOBase.fileno) and half the stuff in [`here`](https://github.com/wikimedia/pywikibot/blame/master/pywikibot/userinterfaces/win32_unicode.py) (I dont recall how much of it is py2 specific, and some aspects wouldnt need mocking).

We can potentially assist with some needs regarding changing of encoding, but it should be separate chunks of work that implement specific requested features.

Reply:
Got it: complex enough that we want to avoid implementing a whole bunch of YAGNI, unless and until specific complaints arise. :+1:

Basically, `stdio-mgr` is not claiming to be a mock for a true *console*...just a mock of the data flow that occurs in common use of `sys.stdfoo`?